src/anomalib/models/image/supersimplenet/torch_model.py

```diff
@@ -0,0 +1,372 @@
+        self.anomaly_generator = SSNAnomalyGenerator(noise_mean=0, noise_std=0.015, threshold=perlin_threshold)
+
+        self.anomaly_map_generator = AnomalyMapGenerator(sigma=4)
+
+    def forward(
+        self,
+        images: torch.Tensor,
+        masks: torch.Tensor | None = None,
+        labels: torch.Tensor | None = None,
+    ) -> InferenceBatch | tuple[torch.Tensor, torch.Tensor, torch.Tensor, torch.Tensor]:
+        """SuperSimpleNet forward pass.
+
+        Extract and process features, adapt them, generate anomalies (train only) and predict anomaly map and score.
+
+        Args:
+            images (torch.Tensor): Input images.
+            masks (torch.Tensor): GT masks.
+            labels (torch.Tensor): GT labels.
+
+        Returns:
+            inference: anomaly map and score
+            training: anomaly map, score and GT masks and labels
+        """
+        output_size = images.shape[-2:]
+
+        features = self.feature_extractor(images)
+        adapted = self.adaptor(features)
+
+        if self.training:
+            masks = self.downsample_mask(masks, *features.shape[-2:])
+            # make linter happy :)
+            if labels is not None:
+                labels = labels.type(torch.float32)
+
+            features, masks, labels = self.anomaly_generator(
+                adapted,
+                masks,
+                labels,
+            )
+
+            anomaly_map, anomaly_score = self.segdec(features)
+            return anomaly_map, anomaly_score, masks, labels
+
+        anomaly_map, anomaly_score = self.segdec(adapted)
+        anomaly_map = self.anomaly_map_generator(anomaly_map, final_size=output_size)
+
+        return InferenceBatch(anomaly_map=anomaly_map, pred_score=anomaly_score)
+
+    @staticmethod
+    def downsample_mask(masks: torch.Tensor, feat_h: int, feat_w: int) -> torch.Tensor:
+        """Downsample the masks according to the feature dimensions.
+
+        Primarily used in supervised setting.
+
+        Args:
+            masks (torch.Tensor): input GT masks
+            feat_h (int): feature height.
+            feat_w (int): feature width.
+
+        Returns:
+            (torch.Tensor): downsampled masks.
+        """
+        masks = masks.type(torch.float32)
+        # best downsampling proposed by DestSeg
+        masks = F.interpolate(
+            masks.unsqueeze(1),
+            size=(feat_h, feat_w),
+            mode="bilinear",
+        )
+        return torch.where(
+            masks < 0.5,
+            torch.zeros_like(masks),
+            torch.ones_like(masks),
+        )
+
+
+def init_weights(m: nn.Module) -> None:
+    """Init weight of the model.
+
+    Args:
+        m (nn.Module): torch module.
+    """
+    if isinstance(m, nn.Linear | nn.Conv2d):
+        nn.init.xavier_normal_(m.weight)
+    elif isinstance(m, nn.BatchNorm1d | nn.BatchNorm2d):
+        nn.init.constant_(m.weight, 1)
+
+
+class FeatureExtractor(nn.Module):
```

Comment:
You could consider cleaning up this file a bit by moving this class and `FeatureAdaptor` to a separate file (e.g. `features.py`). It's up to you, I don't have a strong preference.

Reply:
This is as is since the majority of the models currently follow this structure, so I adhered to it and I think it's best this way.